### PR TITLE
Verify and test

### PR DESCRIPTION
**Pull request by rubyvanrooyen**
Need to get jenkins tests to past, but updates now with integration test scripts and obs_plans

### CODE REVIEW

File: astrokat/observe_main.py [hunk 1]
```diff
@@ -477,7 +475,8 @@ def run_observation(opts, kat):
                         if not tgt:
                             break
                         # check enough time remaining to continue
-                        if time_remaining < tgt['duration']:
+                        if obs_duration > 0 and\
+                           time_remaining < tgt['duration']:
```
Comment (ajoubertza) [inline astrokat/observe_main.py]: Rather use parenthesis, as per our coding guidelines:
```python
if (obs_duration > 0 and
        time_remaining < tgt['duration']):
    done = True
    break
```
Or you could do it all on one line - that would fit within our 90 char limit.

File: astrokat/simulate.py [hunk 2]
```diff
@@ -135,6 +136,7 @@ def __iter__(self):
         raise StopIteration
 
     def __exit__(self, type, value, traceback):
+        # TODO: what is self.track_ used for again?
```
Comment (ajoubertza) [inline astrokat/simulate.py]: This is a bit concerning - would prefer all traces are removed, if there is no purpose.

File: astrokat/simulate.py [hunk 3]
```diff
@@ -135,6 +136,7 @@ def __iter__(self):
...
         if self.track_:
```
Comment (ajoubertza) [inline astrokat/simulate.py]: No longer creating this variable in `__init__`, so can we be sure it will even exist?

File: astrokat/utility.py [hunk 4]
```diff
@@ -17,6 +17,45 @@ class NoTargetsUpError(Exception):
...
+        if data['durations'] is None:
+            msg = 'durations primary key cannot be empty in observation YAML file'
+            raise RuntimeError(msg)
+        if 'start_time' in data['durations']:
+            start_time = data['durations']['start_time']
+            if type(start_time) is str:
+                data['durations']['start_time'] = \
+                        datetime.datetime.strptime(start_time, '%Y-%m-%d %H:%M')
+    if 'observation_loop' not in data.keys():
+        raise RuntimeError('Nothing to observer, exiting')
```
Comment (ajoubertza) [inline astrokat/utility.py]: Typo:  observer -> observe

File: astrokat/utility.py [hunk 5]
```diff
@@ -17,6 +17,45 @@ class NoTargetsUpError(Exception):
...
+    if data['observation_loop'] is None:
+        raise RuntimeError('Empty observation loop, exiting')
+    for obs_loop in data['observation_loop']:
+        if type(obs_loop) is str:
+            raise RuntimeError('Expected observation list, got string')
```
Comment (ajoubertza) [inline astrokat/utility.py]: Error says a list is expected, but the next line uses `obs_loop.keys()`, which suggest it is more like a dict?

File: astrokat/utility.py [hunk 6]
```diff
@@ -17,6 +17,45 @@ class NoTargetsUpError(Exception):
 def read_yaml(filename):
     with open(filename, 'r') as stream:
         data = yaml.safe_load(stream)
+
+    # handle mapping of user friendly keys to CAM resource keys
```
Comment (ajoubertza) [inline astrokat/utility.py]: Better that this is in utils now - thanks.

File: astrokat/observe_main.py (hunk 1 above)
Comment (rubyvanrooyen) [reply astrokat/observe_main.py]: Added the parenthesis implementation , tx

File: astrokat/simulate.py (hunk 2 above)
Comment (rubyvanrooyen) [reply astrokat/simulate.py]: This is for multiple observation loops, needed for zoom modes
I never got time to finish the implementation, sort of like the correlator setup implementation

These are the 2 big hanging implementations that I still need to do and will add separate branches to fix them

File: astrokat/simulate.py (hunk 3 above)
Comment (rubyvanrooyen) [reply astrokat/simulate.py]: Incomplete implementation, only realised the inconsistency when I wrote the planning integration scripts

This must be completed for multiple observation loops, will be addressed in separate devel branch

File: astrokat/utility.py (hunk 5 above)
Comment (rubyvanrooyen) [reply astrokat/utility.py]: Yes, the YAML is read into a JSON style dictionary, but this is for the user input.
In the YAML the observation loops are specified as having an LST range + a list of targets.
Thus, giving an error that is not "quite" correct makes the required input a little clearer.

What about extending the error message to something like?
"
RuntimeError("Incomplete observation input: LST range and list of observations required.")
"

Comment (rubyvanrooyen) [reply astrokat/utility.py]: Suggested implementation
RuntimeError("Incomplete observation input: LST range and a minimum of one target required.")

Comment (rubyvanrooyen) [reply astrokat/utility.py]: Updated error message implemented for more explicit information to user related to possible error.
+
Added TODO note that needs to be implemented and verified with the implementation of multiple observation loop needed for zoom modes.
Issue #47

**Comment by rubyvanrooyen** [comment]
All requested changes have been applied.
Merging to master to unify code

**Pull request merged**